File: modules/collab/test/integration/auth.test.mjs
// Local JWKS harness for the OIDC verifier: generate an RSA keypair, serve its public JWK from a tiny
// local endpoint, sign tokens with the private key, and assert the relay's `authorize` accepts a valid
// token (surfacing the user) and rejects every malformed/expired/wrong-claim one. No real Auth0 tenant,
// no socket — `authorize(req)` is exercised directly with request-shaped objects. Plain ESM (the server
// is .mjs), run by vitest.

import { createServer } from "node:http";
import { SignJWT, exportJWK, generateKeyPair } from "jose";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { createVerifier } from "../../server/auth.mjs";

const ISSUER = "https://test-tenant.example/";
const AUDIENCE = "https://api.mermollusc.test";
const KID = "test-key-1";

let server;
let privateKey;
let authorize;

beforeAll(async () => {
  const { publicKey, privateKey: sk } = await generateKeyPair("RS256");
  privateKey = sk;
  const jwk = { ...(await exportJWK(publicKey)), kid: KID, alg: "RS256", use: "sig" };

  server = createServer((_req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ keys: [jwk] }));
  });
  const port = await new Promise((resolve) => {
    server.listen(0, () => resolve(server.address().port));
  });
  authorize = createVerifier({
    jwksUri: `http://localhost:${port}/.well-known/jwks.json`,
    issuer: ISSUER,
    audience: AUDIENCE,
  });
});

afterAll(() => {
  server.close();
});

const sign = (claims, { issuer = ISSUER, audience = AUDIENCE, expSeconds = 3600 } = {}) =>
  new SignJWT(claims)
    .setProtectedHeader({ alg: "RS256", kid: KID })
    .setIssuer(issuer)
    .setAudience(audience)
    .setIssuedAt()
    .setExpirationTime(Math.floor(Date.now() / 1000) + expSeconds)
    .sign(privateKey);

const req = (token) => (token === null ? { url: "/room" } : { url: "/room", authToken: token });

describe("collab OIDC authorize", () => {
  it("accepts a valid token and surfaces the user (incl. tenant + roles claims)", async () => {
    const token = await sign({
      sub: "auth0|user-123",
      name: "Ada",
      email: "ada@example.com",
      org_id: "org_acme",
      "https://mermollusc.dev/roles": { "org_acme/board1": "editor" },
    });
    const result = await authorize(req(token));
    expect(result.ok).toBe(true);
    expect(result.user).toEqual({
      sub: "auth0|user-123",
      name: "Ada",
      email: "ada@example.com",
      tenant: "org_acme",
      roles: { "org_acme/board1": "editor" },
    });
  });

  it("rejects a missing token", async () => {
    expect((await authorize(req(null))).ok).toBe(false);
  });

  it("rejects a malformed token", async () => {
    expect((await authorize(req("not.a.jwt"))).ok).toBe(false);
  });

  it("rejects a token with the wrong audience", async () => {
    const token = await sign({ sub: "u" }, { audience: "https://some.other.api" });
    expect((await authorize(req(token))).ok).toBe(false);
  });

  it("rejects a token from the wrong issuer", async () => {
    const token = await sign({ sub: "u" }, { issuer: "https://evil.example/" });
    expect((await authorize(req(token))).ok).toBe(false);
  });

  it("rejects an expired token", async () => {
    const token = await sign({ sub: "u" }, { expSeconds: -60 });
    expect((await authorize(req(token))).ok).toBe(false);
  });
});

File: modules/relay/auth/auth_test.go
// Ported from modules/collab/test/integration/auth.test.mjs — a local JWKS harness (generate an RSA
// keypair, serve its public JWK from a tiny local endpoint, sign tokens with the private key) instead of
// a real Auth0 tenant, so the OIDC verification logic is exercised end-to-end without network access.
package auth

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/lestrrat-go/jwx/v3/jwa"
	"github.com/lestrrat-go/jwx/v3/jwk"
	"github.com/lestrrat-go/jwx/v3/jwt"

	"github.com/e6qu/mermollusc/modules/relay/relay"
)

const (
	testIssuer   = "https://test-tenant.example/"
	testAudience = "https://api.mermollusc.test"
	testKID      = "test-key-1"
)

// signingKey bundles the raw RSA private key (unused directly — kept for clarity) with the jwk.Key form
// jwt.Sign needs: signature_builder.go reads the kid to stamp into the JWS header only when the signing
// key itself is a jwk.Key with KeyID set, not a raw *rsa.PrivateKey.
func newTestVerifier(t *testing.T) (relay.Authorizer, jwk.Key) {
	t.Helper()
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	privJWK, err := jwk.Import(priv)
	if err != nil {
		t.Fatalf("import private jwk: %v", err)
	}
	if err := privJWK.Set(jwk.KeyIDKey, testKID); err != nil {
		t.Fatalf("set kid: %v", err)
	}
	if err := privJWK.Set(jwk.AlgorithmKey, jwa.RS256()); err != nil {
		t.Fatalf("set alg: %v", err)
	}

	pub, err := jwk.PublicKeyOf(privJWK)
	if err != nil {
		t.Fatalf("public jwk: %v", err)
	}
	set := jwk.NewSet()
	if err := set.AddKey(pub); err != nil {
		t.Fatalf("add key: %v", err)
	}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(set)
	}))
	t.Cleanup(srv.Close)

	verifier, err := NewVerifier(srv.URL+"/.well-known/jwks.json", testIssuer, testAudience)
	if err != nil {
		t.Fatalf("NewVerifier: %v", err)
	}
	return verifier, privJWK
}

type signOpts struct {
	issuer, audience string
	expIn            time.Duration
}

func sign(t *testing.T, priv jwk.Key, claims map[string]any, opts signOpts) string {
	t.Helper()
	if opts.issuer == "" {
		opts.issuer = testIssuer
	}
	if opts.audience == "" {
		opts.audience = testAudience
	}
	if opts.expIn == 0 {
		opts.expIn = time.Hour
	}
	tok := jwt.New()
	for k, v := range claims {
		if err := tok.Set(k, v); err != nil {
			t.Fatalf("set claim %q: %v", k, err)
		}
	}
	_ = tok.Set(jwt.IssuerKey, opts.issuer)
	_ = tok.Set(jwt.AudienceKey, []string{opts.audience})
	_ = tok.Set(jwt.IssuedAtKey, time.Now())
	_ = tok.Set(jwt.ExpirationKey, time.Now().Add(opts.expIn))

	signed, err := jwt.Sign(tok, jwt.WithKey(jwa.RS256(), priv))
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	return string(signed)
}

func req(t *testing.T, token string) *relay.Request {
	t.Helper()
	return &relay.Request{URL: "/room", AuthToken: token}
}

func TestAcceptsValidTokenAndSurfacesUser(t *testing.T) {
	verifier, priv := newTestVerifier(t)
	token := sign(t, priv, map[string]any{
		"sub":                          "auth0|user-123",
		"name":                         "Ada",
		"email":                        "ada@example.com",
		"org_id":                       "org_acme",
		"https://mermollusc.dev/roles": map[string]any{"org_acme/board1": "editor"},
	}, signOpts{})
	result, err := verifier(context.Background(), req(t, token))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if !result.OK {
		t.Fatalf("result.OK = false, reason=%q", result.Reason)
	}
	want := &relay.User{
		Sub: "auth0|user-123", Name: "Ada", Email: "ada@example.com", Tenant: "org_acme",
		Roles: map[string]relay.Role{"org_acme/board1": relay.RoleEditor},
	}
	if result.User.Sub != want.Sub || result.User.Name != want.Name || result.User.Email != want.Email ||
		result.User.Tenant != want.Tenant || len(result.User.Roles) != 1 ||
		result.User.Roles["org_acme/board1"] != relay.RoleEditor {
		t.Errorf("user = %+v, want %+v", result.User, want)
	}
}

func TestRejectsMissingToken(t *testing.T) {
	verifier, _ := newTestVerifier(t)
	result, err := verifier(context.Background(), req(t, ""))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if result.OK {
		t.Errorf("expected rejection for missing token")
	}
}

func TestRejectsMalformedToken(t *testing.T) {
	verifier, _ := newTestVerifier(t)
	result, err := verifier(context.Background(), req(t, "not.a.jwt"))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if result.OK {
		t.Errorf("expected rejection for malformed token")
	}
}

func TestRejectsWrongAudience(t *testing.T) {
	verifier, priv := newTestVerifier(t)
	token := sign(t, priv, map[string]any{"sub": "u"}, signOpts{audience: "https://some.other.api"})
	result, err := verifier(context.Background(), req(t, token))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if result.OK {
		t.Errorf("expected rejection for wrong audience")
	}
}

func TestRejectsWrongIssuer(t *testing.T) {
	verifier, priv := newTestVerifier(t)
	token := sign(t, priv, map[string]any{"sub": "u"}, signOpts{issuer: "https://evil.example/"})
	result, err := verifier(context.Background(), req(t, token))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if result.OK {
		t.Errorf("expected rejection for wrong issuer")
	}
}

func TestRejectsExpiredToken(t *testing.T) {
	verifier, priv := newTestVerifier(t)
	token := sign(t, priv, map[string]any{"sub": "u"}, signOpts{expIn: -60 * time.Second})
	result, err := verifier(context.Background(), req(t, token))
	if err != nil {
		t.Fatalf("verifier error: %v", err)
	}
	if result.OK {
		t.Errorf("expected rejection for expired token")
	}
}
